jq_properties_parser.go

```diff
@@ -0,0 +1,104 @@
+package opslevel_jq_parser
+
+import (
+	"encoding/json"
+	"strings"
+
+	"github.com/rs/zerolog/log"
+
+	"github.com/opslevel/opslevel-go/v2024"
+)
+
+type JQPropertiesParser struct {
+	assigns []*JQFieldParser
+}
+
+func NewJQPropertiesParser(cfg PropertyRegistrationConfig) *JQPropertiesParser {
+	assigns := make([]*JQFieldParser, len(cfg.Assign))
+	for i, expression := range cfg.Assign {
+		assigns[i] = NewJQFieldParser(expression)
+	}
+	return &JQPropertiesParser{
+		assigns: assigns,
+	}
+}
+
+func propertyAssignmentFromMap(definition string, input map[string]interface{}) (opslevel.PropertyInput, error) {
+	var owner string
+	var value opslevel.JsonString
+	log.Debug().Msgf("property assignment '%s' input is %#v", definition, input)
+	for k, v := range input {
+		switch k {
+		case "owner":
+			owner = v.(string)
+		case "value":
+			jsonString, err := opslevel.NewJSONInput(v)
+			if err != nil {
+				return opslevel.PropertyInput{}, err
+			}
+			value = *jsonString
+		}
+	}
+	if owner != "" && value != "" {
+		return opslevel.PropertyInput{
+			Owner:      *opslevel.NewIdentifier(owner),
+			Definition: *opslevel.NewIdentifier(definition),
+			Value:      value,
+		}, nil
+	}
+	log.Warn().Msgf("got incomplete property assignment '%s'", definition)
+	return opslevel.PropertyInput{}, nil
+}
+
+func (p *JQPropertiesParser) parse(programs []*JQFieldParser, data string) ([]opslevel.PropertyInput, error) {
+	output := make([]opslevel.PropertyInput, 0, len(programs))
+	for _, program := range programs {
+		response, err := program.Run(data)
+		if err != nil {
+			log.Warn().Msg("properties parser: jq error")
+			continue
+		}
+		if response == "" {
+			log.Warn().Msg("properties parser: jq returned empty string")
+			continue
+		}
+
+		if strings.HasPrefix(response, "[") && strings.HasSuffix(response, "]") {
```

Comment:
To reduce the level of nesting in this method could you change this if statement to something like:
```go
if !strings.HasPrefix(response, "[") || !strings.HasSuffix(response, "]") {
    continue
}

// ... with everything currently in if statement moved just below
```
At least for me, it reduces the number of things I need to remember 6 lines down

Reply:
Re-wrote the function with this in mind